test/index.ts

```diff
@@ -1,805 +1,1080 @@
+    genesisBlock.setGenesisParams()
+    blocks.push(...generateBlocks(15, genesisBlock))
+    blockchain.putGenesis(genesisBlock, (err?: Error) => {
+      st.error(err, 'no error')
+      blockchain.putBlocks(blocks.slice(1), (err?: Error) => {
+        if (err) {
+          return st.error(err)
+        }
+        // start: 5, max: 5, skip: 1, reverse: true
+        blockchain.getBlocks(5, 5, 1, true, (err?: any, getBlocks?: any) => {
+          st.error(err, 'no error')
+          st.equal(getBlocks.length, 3)
+          st.ok(getBlocks[0].header.number.equals(blocks[5].header.number))
+          st.ok(!isConsecutive(getBlocks.reverse()), 'blocks should not be consecutive')
+          st.end()
         })
-      },
-      function selectNeededHashes(done) {
+      })
+    })
+  })
+
+  t.test('should find needed hashes', st => {
+    const blockchain = new Blockchain({ validate: false })
+    const blocks: any[] = []
+    const genesisBlock = new Block()
+    genesisBlock.setGenesisParams()
+    blocks.push(...generateBlocks(15, genesisBlock))
+    blockchain.putGenesis(genesisBlock, (err?: Error) => {
+      st.error(err, 'no error')
+      blockchain.putBlocks(blocks.slice(1), (err?: Error) => {
+        if (err) {
+          return st.error(err)
+        }
         const neededHash = Buffer.from('abcdef', 'hex')
         blockchain.selectNeededHashes(
           [blocks[0].hash(), blocks[9].hash(), neededHash],
           (err?: any, hashes?: any) => {
-            if (err) {
-              return done(err)
-            }
-
-            t.equals(
-              hashes[0].toString('hex'),
-              neededHash.toString('hex'),
-              'should find needed hash',
-            )
-            done()
+            st.error(err, 'no error')
+            st.equals(hashes[0].toString('hex'), neededHash.toString('hex'))
+            st.end()
           },
         )
-      },
-      function iterateBlocks(done) {
+      })
+    })
+  })
+
+  t.test('should iterate through 15 blocks', st => {
+    const blockchain = new Blockchain({ validate: false })
+    const blocks: any[] = []
+    const genesisBlock = new Block()
+    genesisBlock.setGenesisParams()
+    blocks.push(...generateBlocks(15, genesisBlock))
+    blockchain.putGenesis(genesisBlock, (err?: Error) => {
+      st.error(err, 'no error')
+      blockchain.putBlocks(blocks.slice(1), (err?: Error) => {
+        if (err) {
+          return st.error(err)
+        }
         let i = 0
         blockchain.iterator(
           'test',
-          function(block: any, _: any, cb: any) {
-            if (block.hash().equals(blocks[i + 1].hash())) i++
+          (block: any, _: any, cb: any) => {
+            if (block.hash().equals(blocks[i + 1].hash())) {
+              i++
+            }
             cb()
           },
-          function() {
-            t.equals(i, 9, 'should iterate through 9 blocks')
-            done()
+          () => {
+            st.equals(i, 15)
```

Comment:
Not sure, but I feel like this should be 14?

Update: nvm realized `generateBlocks` generates n blocks in addition to the genesis blocks.

Reply:
I have changed this a bit now with the new `generateBlockchain` util, should be easier to understand.